// @mailchain/keyring/chunk-CXGOL37P.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/constants.ts
var DERIVATION_PATH_MESSAGING_KEY_ROOT = "Mailchain.Messaging.Root";
var DERIVATION_PATH_IDENTITY_KEY_ROOT = "Mailchain.Identity.Root";
var DERIVATION_PATH_ENCRYPTION_KEY_ROOT = "Mailchain.Encryption.Root";
var DERIVATION_PATH_USER_PROFILE = "UserProfile";
var DERIVATION_PATH_USER_PROFILE_SETTINGS = "UserProfile.Settings";
var DERIVATION_PATH_INBOX_ROOT = "Inbox";
var DERIVATION_PATH_DATE_OFFSET = "DateOffset";









exports.DERIVATION_PATH_MESSAGING_KEY_ROOT = DERIVATION_PATH_MESSAGING_KEY_ROOT; exports.DERIVATION_PATH_IDENTITY_KEY_ROOT = DERIVATION_PATH_IDENTITY_KEY_ROOT; exports.DERIVATION_PATH_ENCRYPTION_KEY_ROOT = DERIVATION_PATH_ENCRYPTION_KEY_ROOT; exports.DERIVATION_PATH_USER_PROFILE = DERIVATION_PATH_USER_PROFILE; exports.DERIVATION_PATH_USER_PROFILE_SETTINGS = DERIVATION_PATH_USER_PROFILE_SETTINGS; exports.DERIVATION_PATH_INBOX_ROOT = DERIVATION_PATH_INBOX_ROOT; exports.DERIVATION_PATH_DATE_OFFSET = DERIVATION_PATH_DATE_OFFSET;
