// @mailchain/internal/chunk-MTKPWAB7.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }// src/sending/payload/create.ts
var _crypto = require('@mailchain/crypto');
var _encoding = require('@mailchain/encoding');
var _isArrayBufferjs = require('lodash/isArrayBuffer.js'); var _isArrayBufferjs2 = _interopRequireDefault(_isArrayBufferjs);
async function createPayload(signerMessagingKey, content, contentType, pluginHeaders) {
  return {
    Headers: {
      Origin: signerMessagingKey.publicKey,
      ContentSignature: await signerMessagingKey.sign(content),
      Created: /* @__PURE__ */ new Date(),
      ContentLength: content.length,
      ContentType: contentType,
      ContentEncoding: _encoding.EncodingTypes.Base64,
      ContentEncryption: _crypto.KindNaClSecretKey,
      PluginHeaders: pluginHeaders
    },
    Content: _isArrayBufferjs2.default.call(void 0, content) ? Buffer.from(content) : content
  };
}



exports.createPayload = createPayload;
