// @mailchain/internal/chunk-YQZCGLMZ.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkO267RXGSjs = require('./chunk-O267RXGS.js');

// src/verifiableCredentials/issuer.ts
var _crypto = require('@mailchain/crypto');
var _encoding = require('@mailchain/encoding');
var _signatures = require('@mailchain/signatures');
var MailchainMessagingKeyIssuer = class {
  async createIssuerFromResolvedAddress(params) {
    const { resolvedAddress, signer } = params;
    if (!_crypto.isPublicKeyEqual.call(void 0, resolvedAddress.messagingKey, signer.publicKey)) {
      return {
        error: new (0, _signatures.ProvidedMessagingKeyIncorrectError)("signer")
      };
    }
    return {
      data: createIssuerFromSigner(resolvedAddress.mailchainAddress, signer)
    };
  }
};
function createIssuerFromSigner(mailchainAddress, signer) {
  return {
    did: _chunkO267RXGSjs.mailchainAddressDecentralizedIdentifier.call(void 0, mailchainAddress),
    alg: "Ed25519",
    signer: async (data) => {
      const dataBytes = typeof data === "string" ? _encoding.decodeUtf8.call(void 0, data) : data;
      const signature = await signer.sign(dataBytes);
      return _encoding.encodeBase64UrlSafe.call(void 0, signature);
    }
  };
}




exports.MailchainMessagingKeyIssuer = MailchainMessagingKeyIssuer; exports.createIssuerFromSigner = createIssuerFromSigner;
