// @mailchain/internal/chunk-XIZIHI7R.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _chunkTRP33P2Rjs = require('./chunk-TRP33P2R.js');


var _chunkIVVOMLHBjs = require('./chunk-IVVOMLHB.js');


var _chunkXXPGZHWZjs = require('./chunk-XXPGZHWZ.js');

// src/sending/distributor/deliveryRequests.ts
var _flattenjs = require('lodash/flatten.js'); var _flattenjs2 = _interopRequireDefault(_flattenjs);
var SendPayloadDistributionRequestsFailuresError = class extends Error {
  constructor(successes, failures) {
    super(
      `Not all mail delivery requests were successfully sent. Check the failed delivery requests to retry failed requests.`
    );
    this.successes = successes;
    this.failures = failures;
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "type", "send_payload_distribution_request_failures");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "docs", "https://docs.mailchain.com/developer/errors/codes#send_payload_distribution_request_failures");
  }
};
var PayloadDeliveryRequests = class {
  constructor(deliveryRequests) {
    this.deliveryRequests = deliveryRequests;
  }
  static create(configuration, sender) {
    return new PayloadDeliveryRequests(_chunkTRP33P2Rjs.DeliveryRequests.create(configuration, sender));
  }
  /**
   * Send the prepared payloads to each recipient.
   * A single payload maybe be sent to multiple recipients in the case of multiple recipients.
   */
  async sendPayloadDistributionRequests(params) {
    const { distributionRequests, resolvedAddresses } = params;
    const sendResults = await Promise.all(
      distributionRequests.map(async (distributionRequest) => {
        const recipients = distributionRequest.distribution.recipients.map(
          (address) => resolvedAddresses.get(address).messagingKey
        );
        const sendManyDeliveryRequestsParams = {
          recipients,
          ...distributionRequest.storedPayload
        };
        const result = await this.deliveryRequests.sendManyDeliveryRequests(sendManyDeliveryRequestsParams);
        return {
          result,
          params: sendManyDeliveryRequestsParams
        };
      })
    );
    const { successes: successfulDeliveries, failures: failed } = _chunkIVVOMLHBjs.partitionMailchainResults.call(void 0, sendResults);
    if (failed.length > 0) {
      return { error: new SendPayloadDistributionRequestsFailuresError(successfulDeliveries, failed) };
    }
    return {
      data: _flattenjs2.default.call(void 0, successfulDeliveries.map(({ data }) => data))
    };
  }
};




exports.SendPayloadDistributionRequestsFailuresError = SendPayloadDistributionRequestsFailuresError; exports.PayloadDeliveryRequests = PayloadDeliveryRequests;
