// @mailchain/internal/chunk-E5ONHO3D.mjs
import {
  __publicField
} from "./chunk-NHABU752.mjs";

// src/messagingKeys/errors.ts
var MessagingKeyContactError = class extends Error {
  constructor(cause) {
    super(`Failed calling messaging key contract.`, { cause });
    __publicField(this, "type", "messaging_key_contact_call_error");
    __publicField(this, "docs", "https://docs.mailchain.com/developer/errors/codes#messaging_key_contact_call_error");
  }
};
var IdentityExpiredError = class extends Error {
  constructor() {
    super(`Identity has expired. Check identity provider settings.`);
    __publicField(this, "type", "identity_expired");
    __publicField(this, "docs", "https://docs.mailchain.com/developer/errors/codes#identity_expired");
  }
};
var IdentityProviderUnsupportedError = class extends Error {
  constructor() {
    super(`Identity provider is not supported. Check list of supported providers.`);
    __publicField(this, "type", "identity_provider_unsupported");
    __publicField(this, "docs", "https://docs.mailchain.com/developer/errors/codes#identity_provider_unsupported");
  }
};
var IdentityNotFoundError = class extends Error {
  constructor() {
    super(`Identity is not found. Check address exists on supplied identity provider.`);
    __publicField(this, "type", "identity_not_found");
    __publicField(this, "docs", "https://docs.mailchain.com/developer/errors/codes#identity_not_found");
  }
};
var IdentityProviderAddressUnsupportedError = class extends Error {
  constructor() {
    super(`Identity is not supported by the provider. Check identity provider supported addresses.`);
    __publicField(this, "type", "identity_provider_address_unsupported");
    __publicField(this, "docs", "https://docs.mailchain.com/developer/errors/codes#identity_provider_address_unsupported");
  }
};
var AddressNotRegisteredError = class extends Error {
  constructor() {
    super("Address is not registered. Register address, then try again.");
    __publicField(this, "type", "address_unregistered");
    __publicField(this, "docs", "https://docs.mailchain.com/developer/errors/codes#address_unregistered");
  }
};

export {
  MessagingKeyContactError,
  IdentityExpiredError,
  IdentityProviderUnsupportedError,
  IdentityNotFoundError,
  IdentityProviderAddressUnsupportedError,
  AddressNotRegisteredError
};
