// @mailchain/internal/chunk-PCGAQDYD.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkXXPGZHWZjs = require('./chunk-XXPGZHWZ.js');

// src/messagingKeys/errors.ts
var MessagingKeyContactError = class extends Error {
  constructor(cause) {
    super(`Failed calling messaging key contract.`, { cause });
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "type", "messaging_key_contact_call_error");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "docs", "https://docs.mailchain.com/developer/errors/codes#messaging_key_contact_call_error");
  }
};
var IdentityExpiredError = class extends Error {
  constructor() {
    super(`Identity has expired. Check identity provider settings.`);
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "type", "identity_expired");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "docs", "https://docs.mailchain.com/developer/errors/codes#identity_expired");
  }
};
var IdentityProviderUnsupportedError = class extends Error {
  constructor() {
    super(`Identity provider is not supported. Check list of supported providers.`);
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "type", "identity_provider_unsupported");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "docs", "https://docs.mailchain.com/developer/errors/codes#identity_provider_unsupported");
  }
};
var IdentityNotFoundError = class extends Error {
  constructor() {
    super(`Identity is not found. Check address exists on supplied identity provider.`);
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "type", "identity_not_found");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "docs", "https://docs.mailchain.com/developer/errors/codes#identity_not_found");
  }
};
var IdentityProviderAddressUnsupportedError = class extends Error {
  constructor() {
    super(`Identity is not supported by the provider. Check identity provider supported addresses.`);
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "type", "identity_provider_address_unsupported");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "docs", "https://docs.mailchain.com/developer/errors/codes#identity_provider_address_unsupported");
  }
};
var AddressNotRegisteredError = class extends Error {
  constructor() {
    super("Address is not registered. Register address, then try again.");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "type", "address_unregistered");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "docs", "https://docs.mailchain.com/developer/errors/codes#address_unregistered");
  }
};








exports.MessagingKeyContactError = MessagingKeyContactError; exports.IdentityExpiredError = IdentityExpiredError; exports.IdentityProviderUnsupportedError = IdentityProviderUnsupportedError; exports.IdentityNotFoundError = IdentityNotFoundError; exports.IdentityProviderAddressUnsupportedError = IdentityProviderAddressUnsupportedError; exports.AddressNotRegisteredError = AddressNotRegisteredError;
