// @mailchain/internal/chunk-AZWVETE6.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }





var _chunkC6QMDU7Djs = require('./chunk-C6QMDU7D.js');


var _chunkM6PQ6TNPjs = require('./chunk-M6PQ6TNP.js');



var _chunkWPXT6MG2js = require('./chunk-WPXT6MG2.js');

// src/transport/payload/headersSerialize.ts
var _canonicalize = require('canonicalize'); var _canonicalize2 = _interopRequireDefault(_canonicalize);
var HEADER_CONTENT_ENCODING = "Content-Encoding";
var HEADER_CONTENT_ENCRYPTION = "Content-Encryption";
var HEADER_CONTENT_LENGTH = "Content-Length";
var HEADER_CONTENT_SIGNATURE = "Content-Signature";
var HEADER_CONTENT_TYPE = "Content-Type";
var HEADER_CREATED = "Created";
var HEADER_ORIGIN = "Origin";
var HEADER_MAILER_CONTENT = "Mailer-Content";
var HEADER_PLUGINS = "Plugin";
var SerializablePayloadHeadersImpl = class {
  deserialize(buffer) {
    const { headers } = _chunkC6QMDU7Djs.headersMapFromBuffers.call(void 0, buffer, [
      HEADER_CONTENT_ENCODING,
      HEADER_CONTENT_ENCRYPTION,
      HEADER_CONTENT_LENGTH,
      HEADER_CONTENT_SIGNATURE,
      HEADER_CONTENT_TYPE,
      HEADER_CREATED,
      HEADER_ORIGIN
    ]);
    const output = {
      ContentEncoding: headers.get(HEADER_CONTENT_ENCODING).toString(),
      ContentEncryption: headers.get(HEADER_CONTENT_ENCRYPTION).toString(),
      ContentLength: Number.parseInt(headers.get(HEADER_CONTENT_LENGTH).toString()),
      ContentSignature: _chunkC6QMDU7Djs.parseSignatureHeader.call(void 0, headers.get(HEADER_CONTENT_SIGNATURE).toString()),
      ContentType: headers.get(HEADER_CONTENT_TYPE).toString(),
      Created: new Date(headers.get(HEADER_CREATED).toString()),
      Origin: _chunkC6QMDU7Djs.parseOriginHeader.call(void 0, headers.get(HEADER_ORIGIN).toString()),
      PluginHeaders: headers.has(HEADER_PLUGINS) ? JSON.parse(headers.get(HEADER_PLUGINS)) : void 0
    };
    if (headers.get(HEADER_MAILER_CONTENT)) {
      const mailerOutput = {
        ...output,
        ContentType: output.ContentType,
        MailerContent: _chunkWPXT6MG2js.parseMailerContentFromJSON.call(void 0, headers.get(HEADER_MAILER_CONTENT).toString())
      };
      return mailerOutput;
    }
    return output;
  }
  serialize(headers) {
    const headersList = [];
    headersList.push(`${HEADER_CONTENT_ENCODING}: ${headers.ContentEncoding}`);
    headersList.push(`${HEADER_CONTENT_ENCRYPTION}: ${headers.ContentEncryption}`);
    headersList.push(`${HEADER_CONTENT_LENGTH}: ${headers.ContentLength}`);
    headersList.push(
      `${HEADER_CONTENT_SIGNATURE}: ${_chunkC6QMDU7Djs.createSignatureHeader.call(void 0, headers.ContentSignature, headers.Origin)}`
    );
    headersList.push(`${HEADER_CONTENT_TYPE}: ${headers.ContentType}`);
    headersList.push(`${HEADER_CREATED}: ${headers.Created.toISOString()}`);
    headersList.push(`${HEADER_ORIGIN}: ${_chunkC6QMDU7Djs.createOriginHeader.call(void 0, headers.Origin)}`);
    if (headers.PluginHeaders) {
      headersList.push(`${HEADER_PLUGINS}: ${_canonicalize2.default.call(void 0, headers.PluginHeaders)}`);
    }
    if (_chunkM6PQ6TNPjs.isResolvedMailerHeaders.call(void 0, headers)) {
      headersList.push(`${HEADER_MAILER_CONTENT}:  ${_chunkWPXT6MG2js.createContentBuffer.call(void 0, headers.MailerContent)}`);
    }
    return Buffer.from(headersList.join("\r\n"), "utf8");
  }
};



exports.SerializablePayloadHeadersImpl = SerializablePayloadHeadersImpl;
