// @mailchain/internal/chunk-EN33MTEW.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _chunk2NXEPH55js = require('./chunk-2NXEPH55.js');

// src/messageSync/messageSync.ts
var _encoding = require('@mailchain/encoding');
var _axios = require('axios'); var _axios2 = _interopRequireDefault(_axios);
var MessageSync = class {
  constructor(sdkConfig, receiverFactory, keyRing, mailboxOperations, axiosInstance = _axios2.default.create()) {
    this.sdkConfig = sdkConfig;
    this.receiverFactory = receiverFactory;
    this.keyRing = keyRing;
    this.mailboxOperations = mailboxOperations;
    this.axiosInstance = axiosInstance;
  }
  static create(sdkConfig, keyRing, mailboxOperations, axiosInstance = _axios2.default.create()) {
    return new MessageSync(sdkConfig, _chunk2NXEPH55js.MailReceiver.create, keyRing, mailboxOperations, axiosInstance);
  }
  async sync(mailboxes) {
    return Promise.all(
      mailboxes.map(
        (mailbox) => this.syncMailbox(mailbox).catch((cause) => ({ status: "fail", mailbox, cause }))
      )
    );
  }
  async syncMailbox(mailbox) {
    const messagingKey = this.keyRing.addressBytesMessagingKey(
      mailbox.messagingKeyParams.address,
      mailbox.messagingKeyParams.protocol,
      mailbox.messagingKeyParams.nonce
    );
    return this.syncWithMessagingKey(mailbox, messagingKey);
  }
  async syncWithMessagingKey(mailbox, messagingKey) {
    const receiver = this.receiverFactory(this.sdkConfig, messagingKey, this.axiosInstance);
    const undeliveredMessages = await receiver.getUndelivered();
    const messages = [];
    for (const undeliveredMessage of undeliveredMessages) {
      if (undeliveredMessage.status !== "success") {
        console.warn("failed to get undelivered message", undeliveredMessage.cause);
        continue;
      }
      const savedMessages = await this.mailboxOperations.saveReceivedMessage({ receivedTransportPayload: undeliveredMessage.payload, userMailbox: mailbox }).catch((e) => {
        console.warn(
          `Failed saving received message with hash ${_encoding.encodeHex.call(void 0, undeliveredMessage.deliveryRequestHash)}`,
          e
        );
        return void 0;
      });
      if (savedMessages && savedMessages.length > 0) {
        messages.push(...savedMessages);
        await receiver.confirmDelivery(undeliveredMessage.deliveryRequestHash).then(() => {
          console.debug(
            `Successfully confirmed delivery message hash ${_encoding.encodeHex.call(void 0, 
              undeliveredMessage.deliveryRequestHash
            )}`
          );
        }).catch(
          (e) => console.warn(
            `Failed saving received message with hash ${_encoding.encodeHex.call(void 0, 
              undeliveredMessage.deliveryRequestHash
            )}`,
            e
          )
        );
      }
    }
    return { status: "success", mailbox, messages };
  }
};



exports.MessageSync = MessageSync;
