// @mailchain/internal/chunk-XFMYRCTF.mjs
// src/mailboxRuleEngine/conditions.ts
function isConditionIsFromAddress(condition) {
  return condition.type === "IsFromAddress";
}
function conditionIsFromAddress(params) {
  return {
    type: "IsFromAddress",
    value: params
  };
}
function isConditionIsFromIdentity(condition) {
  return condition.type === "IsFromIdentity";
}
function conditionIsFromIdentity(params) {
  return {
    type: "IsFromIdentity",
    value: params
  };
}
function conditionOperationOr(conditions) {
  return {
    type: "OperationOr",
    value: conditions
  };
}
function isConditionOperationOr(condition) {
  return condition.type === "OperationOr";
}
function conditionOperationAnd(conditions) {
  return {
    type: "OperationAnd",
    value: conditions
  };
}
function isConditionOperationAnd(condition) {
  return condition.type === "OperationAnd";
}
function conditionOperationNot(condition) {
  return {
    type: "OperationNot",
    value: condition
  };
}
function isConditionOperationNot(condition) {
  return condition.type === "OperationNot";
}

export {
  isConditionIsFromAddress,
  conditionIsFromAddress,
  isConditionIsFromIdentity,
  conditionIsFromIdentity,
  conditionOperationOr,
  isConditionOperationOr,
  conditionOperationAnd,
  isConditionOperationAnd,
  conditionOperationNot,
  isConditionOperationNot
};
