// @mailchain/internal/chunk-EHISW4C7.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/mailboxRuleEngine/conditions.ts
function isConditionIsFromAddress(condition) {
  return condition.type === "IsFromAddress";
}
function conditionIsFromAddress(params) {
  return {
    type: "IsFromAddress",
    value: params
  };
}
function isConditionIsFromIdentity(condition) {
  return condition.type === "IsFromIdentity";
}
function conditionIsFromIdentity(params) {
  return {
    type: "IsFromIdentity",
    value: params
  };
}
function conditionOperationOr(conditions) {
  return {
    type: "OperationOr",
    value: conditions
  };
}
function isConditionOperationOr(condition) {
  return condition.type === "OperationOr";
}
function conditionOperationAnd(conditions) {
  return {
    type: "OperationAnd",
    value: conditions
  };
}
function isConditionOperationAnd(condition) {
  return condition.type === "OperationAnd";
}
function conditionOperationNot(condition) {
  return {
    type: "OperationNot",
    value: condition
  };
}
function isConditionOperationNot(condition) {
  return condition.type === "OperationNot";
}












exports.isConditionIsFromAddress = isConditionIsFromAddress; exports.conditionIsFromAddress = conditionIsFromAddress; exports.isConditionIsFromIdentity = isConditionIsFromIdentity; exports.conditionIsFromIdentity = conditionIsFromIdentity; exports.conditionOperationOr = conditionOperationOr; exports.isConditionOperationOr = isConditionOperationOr; exports.conditionOperationAnd = conditionOperationAnd; exports.isConditionOperationAnd = isConditionOperationAnd; exports.conditionOperationNot = conditionOperationNot; exports.isConditionOperationNot = isConditionOperationNot;
