// intercom-client/api/resources/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.visitors = exports.ticketTypes = exports.phoneCallRedirects = exports.events = exports.helpCenters = exports.news = exports.dataEvents = exports.customObjectInstances = exports.aiContentSource = exports.helpCenter = exports.aiAgent = exports.tickets = exports.teams = exports.subscriptionTypes = exports.segments = exports.messages = exports.dataExport = exports.dataAttributes = exports.conversations = exports.tags = exports.notes = exports.contacts = exports.companies = exports.articles = exports.admins = void 0;
exports.admins = __importStar(require("./admins"));
__exportStar(require("./admins/types"), exports);
exports.articles = __importStar(require("./articles"));
__exportStar(require("./articles/types"), exports);
exports.companies = __importStar(require("./companies"));
__exportStar(require("./companies/types"), exports);
exports.contacts = __importStar(require("./contacts"));
__exportStar(require("./contacts/types"), exports);
exports.notes = __importStar(require("./notes"));
__exportStar(require("./notes/types"), exports);
exports.tags = __importStar(require("./tags"));
__exportStar(require("./tags/types"), exports);
exports.conversations = __importStar(require("./conversations"));
__exportStar(require("./conversations/types"), exports);
exports.dataAttributes = __importStar(require("./dataAttributes"));
__exportStar(require("./dataAttributes/types"), exports);
exports.dataExport = __importStar(require("./dataExport"));
__exportStar(require("./dataExport/types"), exports);
exports.messages = __importStar(require("./messages"));
__exportStar(require("./messages/types"), exports);
exports.segments = __importStar(require("./segments"));
__exportStar(require("./segments/types"), exports);
exports.subscriptionTypes = __importStar(require("./subscriptionTypes"));
__exportStar(require("./subscriptionTypes/types"), exports);
exports.teams = __importStar(require("./teams"));
__exportStar(require("./teams/types"), exports);
exports.tickets = __importStar(require("./tickets"));
__exportStar(require("./tickets/types"), exports);
exports.aiAgent = __importStar(require("./aiAgent"));
__exportStar(require("./aiAgent/types"), exports);
exports.helpCenter = __importStar(require("./helpCenter"));
__exportStar(require("./helpCenter/types"), exports);
exports.aiContentSource = __importStar(require("./aiContentSource"));
__exportStar(require("./aiContentSource/types"), exports);
exports.customObjectInstances = __importStar(require("./customObjectInstances"));
__exportStar(require("./customObjectInstances/types"), exports);
exports.dataEvents = __importStar(require("./dataEvents"));
__exportStar(require("./dataEvents/types"), exports);
exports.news = __importStar(require("./news"));
__exportStar(require("./news/types"), exports);
exports.helpCenters = __importStar(require("./helpCenters"));
exports.events = __importStar(require("./events"));
exports.phoneCallRedirects = __importStar(require("./phoneCallRedirects"));
exports.ticketTypes = __importStar(require("./ticketTypes"));
exports.visitors = __importStar(require("./visitors"));
__exportStar(require("./admins/client/requests"), exports);
__exportStar(require("./articles/client/requests"), exports);
__exportStar(require("./helpCenters/client/requests"), exports);
__exportStar(require("./companies/client/requests"), exports);
__exportStar(require("./contacts/client/requests"), exports);
__exportStar(require("./notes/client/requests"), exports);
__exportStar(require("./tags/client/requests"), exports);
__exportStar(require("./conversations/client/requests"), exports);
__exportStar(require("./dataAttributes/client/requests"), exports);
__exportStar(require("./events/client/requests"), exports);
__exportStar(require("./dataExport/client/requests"), exports);
__exportStar(require("./segments/client/requests"), exports);
__exportStar(require("./phoneCallRedirects/client/requests"), exports);
__exportStar(require("./teams/client/requests"), exports);
__exportStar(require("./ticketTypes/client/requests"), exports);
__exportStar(require("./tickets/client/requests"), exports);
__exportStar(require("./visitors/client/requests"), exports);
