// @mailchain/api/chunk-7VCGQ36W.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _chunkXXPGZHWZjs = require('./chunk-XXPGZHWZ.js');

// src/api/base.ts
var _axios = require('axios'); var _axios2 = _interopRequireDefault(_axios);
var BASE_PATH = "http://localhost".replace(/\/+$/, "");
var COLLECTION_FORMATS = {
  csv: ",",
  ssv: " ",
  tsv: "	",
  pipes: "|"
};
var BaseAPI = class {
  constructor(configuration, basePath = BASE_PATH, axios = _axios2.default) {
    this.basePath = basePath;
    this.axios = axios;
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "configuration");
    if (configuration) {
      this.configuration = configuration;
      this.basePath = configuration.basePath || this.basePath;
    }
  }
};
var RequiredError = class extends Error {
  constructor(field, msg) {
    super(msg);
    this.field = field;
    this.name = "RequiredError";
  }
};






exports.BASE_PATH = BASE_PATH; exports.COLLECTION_FORMATS = COLLECTION_FORMATS; exports.BaseAPI = BaseAPI; exports.RequiredError = RequiredError;
