// intercom-client/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntercomClient = void 0;
const Client_1 = require("./api/resources/admins/client/Client");
const Client_2 = require("./api/resources/articles/client/Client");
const Client_3 = require("./api/resources/helpCenters/client/Client");
const Client_4 = require("./api/resources/companies/client/Client");
const Client_5 = require("./api/resources/contacts/client/Client");
const Client_6 = require("./api/resources/notes/client/Client");
const Client_7 = require("./api/resources/tags/client/Client");
const Client_8 = require("./api/resources/conversations/client/Client");
const Client_9 = require("./api/resources/dataAttributes/client/Client");
const Client_10 = require("./api/resources/events/client/Client");
const Client_11 = require("./api/resources/dataExport/client/Client");
const Client_12 = require("./api/resources/messages/client/Client");
const Client_13 = require("./api/resources/segments/client/Client");
const Client_14 = require("./api/resources/subscriptionTypes/client/Client");
const Client_15 = require("./api/resources/phoneCallRedirects/client/Client");
const Client_16 = require("./api/resources/teams/client/Client");
const Client_17 = require("./api/resources/ticketTypes/client/Client");
const Client_18 = require("./api/resources/tickets/client/Client");
const Client_19 = require("./api/resources/visitors/client/Client");
const Client_20 = require("./api/resources/news/client/Client");
class IntercomClient {
    constructor(_options = {}) {
        this._options = _options;
    }
    get admins() {
        var _a;
        return ((_a = this._admins) !== null && _a !== void 0 ? _a : (this._admins = new Client_1.Admins(this._options)));
    }
    get articles() {
        var _a;
        return ((_a = this._articles) !== null && _a !== void 0 ? _a : (this._articles = new Client_2.Articles(this._options)));
    }
    get helpCenters() {
        var _a;
        return ((_a = this._helpCenters) !== null && _a !== void 0 ? _a : (this._helpCenters = new Client_3.HelpCenters(this._options)));
    }
    get companies() {
        var _a;
        return ((_a = this._companies) !== null && _a !== void 0 ? _a : (this._companies = new Client_4.Companies(this._options)));
    }
    get contacts() {
        var _a;
        return ((_a = this._contacts) !== null && _a !== void 0 ? _a : (this._contacts = new Client_5.Contacts(this._options)));
    }
    get notes() {
        var _a;
        return ((_a = this._notes) !== null && _a !== void 0 ? _a : (this._notes = new Client_6.Notes(this._options)));
    }
    get tags() {
        var _a;
        return ((_a = this._tags) !== null && _a !== void 0 ? _a : (this._tags = new Client_7.Tags(this._options)));
    }
    get conversations() {
        var _a;
        return ((_a = this._conversations) !== null && _a !== void 0 ? _a : (this._conversations = new Client_8.Conversations(this._options)));
    }
    get dataAttributes() {
        var _a;
        return ((_a = this._dataAttributes) !== null && _a !== void 0 ? _a : (this._dataAttributes = new Client_9.DataAttributes(this._options)));
    }
    get events() {
        var _a;
        return ((_a = this._events) !== null && _a !== void 0 ? _a : (this._events = new Client_10.Events(this._options)));
    }
    get dataExport() {
        var _a;
        return ((_a = this._dataExport) !== null && _a !== void 0 ? _a : (this._dataExport = new Client_11.DataExport(this._options)));
    }
    get messages() {
        var _a;
        return ((_a = this._messages) !== null && _a !== void 0 ? _a : (this._messages = new Client_12.Messages(this._options)));
    }
    get segments() {
        var _a;
        return ((_a = this._segments) !== null && _a !== void 0 ? _a : (this._segments = new Client_13.Segments(this._options)));
    }
    get subscriptionTypes() {
        var _a;
        return ((_a = this._subscriptionTypes) !== null && _a !== void 0 ? _a : (this._subscriptionTypes = new Client_14.SubscriptionTypes(this._options)));
    }
    get phoneCallRedirects() {
        var _a;
        return ((_a = this._phoneCallRedirects) !== null && _a !== void 0 ? _a : (this._phoneCallRedirects = new Client_15.PhoneCallRedirects(this._options)));
    }
    get teams() {
        var _a;
        return ((_a = this._teams) !== null && _a !== void 0 ? _a : (this._teams = new Client_16.Teams(this._options)));
    }
    get ticketTypes() {
        var _a;
        return ((_a = this._ticketTypes) !== null && _a !== void 0 ? _a : (this._ticketTypes = new Client_17.TicketTypes(this._options)));
    }
    get tickets() {
        var _a;
        return ((_a = this._tickets) !== null && _a !== void 0 ? _a : (this._tickets = new Client_18.Tickets(this._options)));
    }
    get visitors() {
        var _a;
        return ((_a = this._visitors) !== null && _a !== void 0 ? _a : (this._visitors = new Client_19.Visitors(this._options)));
    }
    get news() {
        var _a;
        return ((_a = this._news) !== null && _a !== void 0 ? _a : (this._news = new Client_20.News(this._options)));
    }
}
exports.IntercomClient = IntercomClient;
