// @mailchain/sdk/internal/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunk2JM2DQYVjs = require('../chunk-2JM2DQYV.js');





var _chunkYSZOFUIZjs = require('../chunk-YSZOFUIZ.js');


var _chunkMWOVQ7KAjs = require('../chunk-MWOVQ7KA.js');


var _chunkGNQZPSQSjs = require('../chunk-GNQZPSQS.js');
require('../chunk-XXPGZHWZ.js');

// src/internal/index.ts
var _mail = require('@mailchain/internal/sending/mail');
var _verifiableCredentials = require('@mailchain/internal/verifiableCredentials');
var _verifiablePresentationRequest = require('@mailchain/internal/sending/verifiablePresentationRequest');











exports.MailSender = _mail.MailSender; exports.MailchainAddressOwnershipVerifier = _verifiableCredentials.MailchainAddressOwnershipVerifier; exports.VerifiablePresentationRequestSender = _verifiablePresentationRequest.VerifiablePresentationRequestSender; exports.getMessagingKeyLatestNonce = _chunk2JM2DQYVjs.getMessagingKeyLatestNonce; exports.getPrivateMessagingKey = _chunkMWOVQ7KAjs.getPrivateMessagingKey; exports.privateMessagingKeyFromHex = _chunkYSZOFUIZjs.privateMessagingKeyFromHex; exports.privateMessagingKeyToHex = _chunkYSZOFUIZjs.privateMessagingKeyToHex; exports.publicMessagingKeyFromHex = _chunkYSZOFUIZjs.publicMessagingKeyFromHex; exports.publicMessagingKeyToHex = _chunkYSZOFUIZjs.publicMessagingKeyToHex; exports.resolveAddress = _chunkGNQZPSQSjs.resolveAddress;
