// intercom-client/api/types/ActivityLog.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityLog = void 0;
var ActivityLog;
(function (ActivityLog) {
    ActivityLog.ActivityType = {
        AdminAssignmentLimitChange: "admin_assignment_limit_change",
        AdminAwayModeChange: "admin_away_mode_change",
        AdminDeletion: "admin_deletion",
        AdminDeprovisioned: "admin_deprovisioned",
        AdminImpersonationEnd: "admin_impersonation_end",
        AdminImpersonationStart: "admin_impersonation_start",
        AdminInviteChange: "admin_invite_change",
        AdminInviteCreation: "admin_invite_creation",
        AdminInviteDeletion: "admin_invite_deletion",
        AdminLoginFailure: "admin_login_failure",
        AdminLoginSuccess: "admin_login_success",
        AdminLogout: "admin_logout",
        AdminPasswordResetRequest: "admin_password_reset_request",
        AdminPasswordResetSuccess: "admin_password_reset_success",
        AdminPermissionChange: "admin_permission_change",
        AdminProvisioned: "admin_provisioned",
        AdminTwoFactorAuthChange: "admin_two_factor_auth_change",
        AdminUnauthorizedSignInMethod: "admin_unauthorized_sign_in_method",
        AppAdminJoin: "app_admin_join",
        AppAuthenticationMethodChange: "app_authentication_method_change",
        AppDataDeletion: "app_data_deletion",
        AppDataExport: "app_data_export",
        AppGoogleSsoDomainChange: "app_google_sso_domain_change",
        AppIdentityVerificationChange: "app_identity_verification_change",
        AppNameChange: "app_name_change",
        AppOutboundAddressChange: "app_outbound_address_change",
        AppPackageInstallation: "app_package_installation",
        AppPackageTokenRegeneration: "app_package_token_regeneration",
        AppPackageUninstallation: "app_package_uninstallation",
        AppTeamCreation: "app_team_creation",
        AppTeamDeletion: "app_team_deletion",
        AppTeamMembershipModification: "app_team_membership_modification",
        AppTimezoneChange: "app_timezone_change",
        AppWebhookCreation: "app_webhook_creation",
        AppWebhookDeletion: "app_webhook_deletion",
        ArticlesInMessengerEnabledChange: "articles_in_messenger_enabled_change",
        BulkDelete: "bulk_delete",
        BulkExport: "bulk_export",
        CampaignDeletion: "campaign_deletion",
        CampaignStateChange: "campaign_state_change",
        ConversationPartDeletion: "conversation_part_deletion",
        ConversationTopicChange: "conversation_topic_change",
        ConversationTopicCreation: "conversation_topic_creation",
        ConversationTopicDeletion: "conversation_topic_deletion",
        HelpCenterSettingsChange: "help_center_settings_change",
        InboundConversationsChange: "inbound_conversations_change",
        InboxAccessChange: "inbox_access_change",
        MessageDeletion: "message_deletion",
        MessageStateChange: "message_state_change",
        MessengerLookAndFeelChange: "messenger_look_and_feel_change",
        MessengerSearchRequiredChange: "messenger_search_required_change",
        MessengerSpacesChange: "messenger_spaces_change",
        OfficeHoursChange: "office_hours_change",
        RoleChange: "role_change",
        RoleCreation: "role_creation",
        RoleDeletion: "role_deletion",
        RulesetActivationTitlePreview: "ruleset_activation_title_preview",
        RulesetCreation: "ruleset_creation",
        RulesetDeletion: "ruleset_deletion",
        SearchBrowseEnabledChange: "search_browse_enabled_change",
        SearchBrowseRequiredChange: "search_browse_required_change",
        SeatChange: "seat_change",
        SeatRevoke: "seat_revoke",
        SecuritySettingsChange: "security_settings_change",
        TemporaryExpectationChange: "temporary_expectation_change",
        UpfrontEmailCollectionChange: "upfront_email_collection_change",
        WelcomeMessageChange: "welcome_message_change",
    };
})(ActivityLog = exports.ActivityLog || (exports.ActivityLog = {}));
