// @mailchain/internal/transport/mailer/author.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});require('../../chunk-HXC7OVSM.js');


var _chunk5FSOOI2Gjs = require('../../chunk-5FSOOI2G.js');
require('../../chunk-VRWBLL7X.js');
require('../../chunk-ILVQWVUY.js');
require('../../chunk-ESCCOX3V.js');
require('../../chunk-7SVBKWIY.js');
require('../../chunk-FJ3BB6CV.js');
require('../../chunk-BQQXYBRH.js');
require('../../chunk-DIK6R3YN.js');
require('../../chunk-7K4JOLRR.js');
require('../../chunk-AR4N7KS5.js');
require('../../chunk-GDYJBQYG.js');
require('../../chunk-PCGAQDYD.js');
require('../../chunk-IVVOMLHB.js');
require('../../chunk-ZFRYLPIO.js');
require('../../chunk-RH2527U6.js');
require('../../chunk-XXPGZHWZ.js');

// src/transport/mailer/author.ts
var _signatures = require('@mailchain/signatures');
var MailerAuthorVerifier = class {
  constructor(sender) {
    this.sender = sender;
  }
  static create(configuration) {
    return new MailerAuthorVerifier(_chunk5FSOOI2Gjs.SenderVerifier.create(configuration));
  }
  async verifyMailerAuthor(mailerContent) {
    if (!this.verifyMailerProof(mailerContent.authorMessagingKey, mailerContent.mailerProof)) {
      throw new Error("mailer proof is not valid");
    }
    if (!this.verifyAuthorWroteContent(
      mailerContent.contentUri,
      mailerContent.mailerProof.params.authorContentSignature,
      mailerContent.authorMessagingKey
    )) {
      throw new Error("authorContentSignature is not valid");
    }
    if (!this.verifyAuthorOwnsFromAddress(mailerContent.authorMailAddress, mailerContent.authorMessagingKey)) {
      throw new Error("messaging is not the latest message key for author address");
    }
    return;
  }
  async verifyAuthorWroteContent(contentUri, authorContentSignature, authorMessagingKey) {
    return authorMessagingKey.verify(Buffer.from(contentUri), authorContentSignature);
  }
  async verifyAuthorOwnsFromAddress(fromAddress, authorMessagingKey) {
    return await this.sender.verifySenderOwnsFromAddress(fromAddress.address, authorMessagingKey);
  }
  async verifyMailerProof(authorMessagingKey, mailerProof) {
    return _signatures.verifyMailerProof.call(void 0, authorMessagingKey, mailerProof);
  }
};


exports.MailerAuthorVerifier = MailerAuthorVerifier;
