// @mailchain/internal/transport/serialization/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});require('../../chunk-GICIHBQY.js');



var _chunkIXZMJHSYjs = require('../../chunk-IXZMJHSY.js');


var _chunkRQT6CC6Tjs = require('../../chunk-RQT6CC6T.js');


var _chunkI6LRDV5Yjs = require('../../chunk-I6LRDV5Y.js');


var _chunkVPET7X4Njs = require('../../chunk-VPET7X4N.js');







var _chunkC6QMDU7Djs = require('../../chunk-C6QMDU7D.js');
require('../../chunk-XXPGZHWZ.js');












exports.CHUNK_LENGTH_1MB = _chunkVPET7X4Njs.CHUNK_LENGTH_1MB; exports.createOriginHeader = _chunkC6QMDU7Djs.createOriginHeader; exports.createSignatureHeader = _chunkC6QMDU7Djs.createSignatureHeader; exports.decryptPayload = _chunkRQT6CC6Tjs.decryptPayload; exports.deserialize = _chunkIXZMJHSYjs.deserialize; exports.encryptPayload = _chunkI6LRDV5Yjs.encryptPayload; exports.headersMapFromBuffers = _chunkC6QMDU7Djs.headersMapFromBuffers; exports.parseHeaderElements = _chunkC6QMDU7Djs.parseHeaderElements; exports.parseOriginHeader = _chunkC6QMDU7Djs.parseOriginHeader; exports.parseSignatureHeader = _chunkC6QMDU7Djs.parseSignatureHeader; exports.serialize = _chunkIXZMJHSYjs.serialize;
