// @mailchain/crypto/chunk-B5NCYBFK.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkEMN3YRZAjs = require('./chunk-EMN3YRZA.js');

// src/scrypt/scrypt.ts
var _scrypt = require('@noble/hashes/scrypt');
var defaultScryptParams = {
  // N is the N parameter of Scrypt encryption algorithm, using 256MB
  // memory and taking approximately 1s CPU time on a modern processor.
  N: 1 << 18,
  // P is the P parameter of Scrypt encryption algorithm, using 256MB
  // memory and taking approximately 1s CPU time on a modern processor.
  p: 1,
  r: 8,
  keyLen: 32
};
function deriveSecretFromScrypt(passphrase = "", params = defaultScryptParams, salt = _chunkEMN3YRZAjs.secureRandom.call(void 0, 32)) {
  const secret = _scrypt.scrypt.call(void 0, new Uint8Array(Buffer.from(passphrase, "ascii")), salt, {
    N: params.N,
    r: params.r,
    p: params.p,
    dkLen: params.keyLen
  });
  return {
    params,
    secret,
    salt
  };
}




exports.defaultScryptParams = defaultScryptParams; exports.deriveSecretFromScrypt = deriveSecretFromScrypt;
