// @mailchain/crypto/chunk-BSMIECIG.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/keys.ts
var KindSECP256K1 = "secp256k1";
var KindED25519 = "ed25519";
var KindSECP256R1 = "secp256r1";
var KindSR25519 = "sr25519";
var KindNoOp = "noop";
var IdSECP256K1 = 225;
var IdED25519 = 226;
var IdSR25519 = 227;
var IdSECP256R1 = 228;
var IdNonSpecified = 238;
var CurveIds = /* @__PURE__ */ ((CurveIds2) => {
  CurveIds2[CurveIds2["SECP256K1"] = IdSECP256K1] = "SECP256K1";
  CurveIds2[CurveIds2["ED25519"] = IdED25519] = "ED25519";
  CurveIds2[CurveIds2["SR25519"] = IdSR25519] = "SR25519";
  CurveIds2[CurveIds2["SECP256R1"] = IdSECP256R1] = "SECP256R1";
  return CurveIds2;
})(CurveIds || {});













exports.KindSECP256K1 = KindSECP256K1; exports.KindED25519 = KindED25519; exports.KindSECP256R1 = KindSECP256R1; exports.KindSR25519 = KindSR25519; exports.KindNoOp = KindNoOp; exports.IdSECP256K1 = IdSECP256K1; exports.IdED25519 = IdED25519; exports.IdSR25519 = IdSR25519; exports.IdSECP256R1 = IdSECP256R1; exports.IdNonSpecified = IdNonSpecified; exports.CurveIds = CurveIds;
