// @mailchain/crypto/chunk-FKBKPFXU.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _chunkEMN3YRZAjs = require('./chunk-EMN3YRZA.js');


var _chunkHWVQHADCjs = require('./chunk-HWVQHADC.js');


var _chunkBSMIECIGjs = require('./chunk-BSMIECIG.js');


var _chunkGJHEVIDSjs = require('./chunk-GJHEVIDS.js');


var _chunkXXPGZHWZjs = require('./chunk-XXPGZHWZ.js');

// src/ed25519/private.ts
var _tweetnacl = require('tweetnacl'); var _tweetnacl2 = _interopRequireDefault(_tweetnacl);
var ED25519PrivateKeyLength = 64;
var ED25519SeedLength = 32;
var ED25519PrivateKey = class {
  constructor(keyPair) {
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "bytes");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "publicKey");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "curve", _chunkBSMIECIGjs.KindED25519);
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "keyPair");
    this.keyPair = keyPair;
    this.bytes = this.keyPair.secretKey;
    this.publicKey = new (0, _chunkHWVQHADCjs.ED25519PublicKey)(this.keyPair.publicKey);
  }
  static fromSeed(seed) {
    if (seed.length !== ED25519SeedLength) {
      throw Error("seed must be 32 bytes");
    }
    return new this(_tweetnacl2.default.sign.keyPair.fromSeed(seed));
  }
  static fromSecretKey(secretKey) {
    if (secretKey.length !== ED25519PrivateKeyLength) {
      throw Error("secret key must be 64 bytes");
    }
    return new this(_tweetnacl2.default.sign.keyPair.fromSecretKey(secretKey));
  }
  static fromMnemonicPhrase(mnemonic, password = "") {
    return this.fromSeed(_chunkGJHEVIDSjs.toSeed.call(void 0, mnemonic, password, ED25519SeedLength));
  }
  static generate(rand = _chunkEMN3YRZAjs.secureRandom) {
    return this.fromSeed(rand(ED25519SeedLength));
  }
  async sign(message) {
    return _tweetnacl2.default.sign.detached(message, this.keyPair.secretKey);
  }
};
function asED25519PrivateKey(key) {
  if (key.constructor !== ED25519PrivateKey) {
    throw new Error("key must be ed25519");
  }
  return key;
}






exports.ED25519PrivateKeyLength = ED25519PrivateKeyLength; exports.ED25519SeedLength = ED25519SeedLength; exports.ED25519PrivateKey = ED25519PrivateKey; exports.asED25519PrivateKey = asED25519PrivateKey;
