// @mailchain/addressing/addressPredicates.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});







var _chunkKULIOUHZjs = require('./chunk-KULIOUHZ.js');
require('./chunk-E6OAXMPT.js');
require('./chunk-OS5AHDT5.js');
require('./chunk-L7UXQ4P2.js');
require('./chunk-BAZMUWY4.js');
require('./chunk-PR4HBQYQ.js');
require('./chunk-PV3ICE6E.js');
require('./chunk-XESCJ3QG.js');
require('./chunk-P66WQYVA.js');
require('./chunk-7QWM3DS3.js');
require('./chunk-O73NCH4N.js');
require('./chunk-QBWJRO3P.js');
require('./chunk-MFGUARH3.js');
require('./chunk-7K2RPUW3.js');
require('./chunk-6IKXH2YO.js');
require('./chunk-XXPGZHWZ.js');








exports.isEthereumAddress = _chunkKULIOUHZjs.isEthereumAddress; exports.isFilecoinAddress = _chunkKULIOUHZjs.isFilecoinAddress; exports.isMailchainAccountAddress = _chunkKULIOUHZjs.isMailchainAccountAddress; exports.isNearImplicitAccount = _chunkKULIOUHZjs.isNearImplicitAccount; exports.isSolanaAddress = _chunkKULIOUHZjs.isSolanaAddress; exports.isTezosAddress = _chunkKULIOUHZjs.isTezosAddress; exports.isTokenAddress = _chunkKULIOUHZjs.isTokenAddress;
