// @mailchain/addressing/chunk-SSBZ77W5.mjs
// src/protocols/filecoin/const.ts
import { SECP256K1PrivateKey } from "@mailchain/crypto";
import { decodeHex } from "@mailchain/encoding";
var AliceFilStr = "f410faiikgixcdd7alrsbpunjlen4etkcarbfbghn7wi";
var AliceFilEthAddressStr = "0x0210a322e218fe05c6417d1a9591bc24d4204425";
var AliceFilEthPrivateKeyStr = "6c115551accbd74b46401a531af1dd0890034c18137dddab52a53c626e75b051";
var AliceFilEthPrivateKey = new SECP256K1PrivateKey(decodeHex(AliceFilEthPrivateKeyStr));
var BobFilAddressStr = "f410f5blfdi7tmto6o2aa2up4vwtdm7qcnaiqpeazuwy";
var BobFilEthAddressStr = "0xe85651a3f364dde76800d51fcada6367e0268110";
var BobFilEthPrivateKeyStr = "2c6b29d5545e5804b7fb7e65404782e57dddea84cbcec9b2586add6403808965";
var BobFilEthPrivateKey = new SECP256K1PrivateKey(decodeHex(BobFilEthPrivateKeyStr));
var FILECOIN_PREFIX_F4_ETHEREUM = "f410";
var FILECOIN_PREFIX_T4_ETHEREUM = "t410";

export {
  AliceFilStr,
  AliceFilEthAddressStr,
  AliceFilEthPrivateKeyStr,
  AliceFilEthPrivateKey,
  BobFilAddressStr,
  BobFilEthAddressStr,
  BobFilEthPrivateKeyStr,
  BobFilEthPrivateKey,
  FILECOIN_PREFIX_F4_ETHEREUM,
  FILECOIN_PREFIX_T4_ETHEREUM
};
