// @mailchain/addressing/chunk-B46YFNHS.mjs
// src/protocols/tezos/const.ts
var Prefix = /* @__PURE__ */ ((Prefix2) => {
  Prefix2["TZ1"] = "tz1";
  Prefix2["TZ2"] = "tz2";
  Prefix2["TZ3"] = "tz3";
  Prefix2["KT"] = "KT";
  Prefix2["KT1"] = "KT1";
  Prefix2["EDSK2"] = "edsk2";
  Prefix2["SPSK"] = "spsk";
  Prefix2["P2SK"] = "p2sk";
  Prefix2["EDPK"] = "edpk";
  Prefix2["SPPK"] = "sppk";
  Prefix2["P2PK"] = "p2pk";
  Prefix2["EDESK"] = "edesk";
  Prefix2["SPESK"] = "spesk";
  Prefix2["P2ESK"] = "p2esk";
  Prefix2["EDSK"] = "edsk";
  Prefix2["EDSIG"] = "edsig";
  Prefix2["SPSIG"] = "spsig";
  Prefix2["P2SIG"] = "p2sig";
  Prefix2["SIG"] = "sig";
  Prefix2["EXPR"] = "expr";
  return Prefix2;
})(Prefix || {});
var prefix = {
  ["tz1" /* TZ1 */]: new Uint8Array([6, 161, 159]),
  ["tz2" /* TZ2 */]: new Uint8Array([6, 161, 161]),
  ["tz3" /* TZ3 */]: new Uint8Array([6, 161, 164]),
  ["KT" /* KT */]: new Uint8Array([2, 90, 121]),
  ["KT1" /* KT1 */]: new Uint8Array([2, 90, 121]),
  ["edsk" /* EDSK */]: new Uint8Array([43, 246, 78, 7]),
  ["edsk2" /* EDSK2 */]: new Uint8Array([13, 15, 58, 7]),
  ["spsk" /* SPSK */]: new Uint8Array([17, 162, 224, 201]),
  ["p2sk" /* P2SK */]: new Uint8Array([16, 81, 238, 189]),
  ["edpk" /* EDPK */]: new Uint8Array([13, 15, 37, 217]),
  ["sppk" /* SPPK */]: new Uint8Array([3, 254, 226, 86]),
  ["p2pk" /* P2PK */]: new Uint8Array([3, 178, 139, 127]),
  ["edesk" /* EDESK */]: new Uint8Array([7, 90, 60, 179, 41]),
  ["spesk" /* SPESK */]: new Uint8Array([9, 237, 241, 174, 150]),
  ["p2esk" /* P2ESK */]: new Uint8Array([9, 48, 57, 115, 171]),
  ["edsig" /* EDSIG */]: new Uint8Array([9, 245, 205, 134, 18]),
  ["spsig" /* SPSIG */]: new Uint8Array([13, 115, 101, 19, 63]),
  ["p2sig" /* P2SIG */]: new Uint8Array([54, 240, 44, 52]),
  ["sig" /* SIG */]: new Uint8Array([4, 130, 43]),
  ["expr" /* EXPR */]: new Uint8Array([13, 44, 64, 27])
};

export {
  Prefix,
  prefix
};
