// @mailchain/addressing/chunk-PR4HBQYQ.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/protocols/filecoin/const.ts
var _crypto = require('@mailchain/crypto');
var _encoding = require('@mailchain/encoding');
var AliceFilStr = "f410faiikgixcdd7alrsbpunjlen4etkcarbfbghn7wi";
var AliceFilEthAddressStr = "0x0210a322e218fe05c6417d1a9591bc24d4204425";
var AliceFilEthPrivateKeyStr = "6c115551accbd74b46401a531af1dd0890034c18137dddab52a53c626e75b051";
var AliceFilEthPrivateKey = new (0, _crypto.SECP256K1PrivateKey)(_encoding.decodeHex.call(void 0, AliceFilEthPrivateKeyStr));
var BobFilAddressStr = "f410f5blfdi7tmto6o2aa2up4vwtdm7qcnaiqpeazuwy";
var BobFilEthAddressStr = "0xe85651a3f364dde76800d51fcada6367e0268110";
var BobFilEthPrivateKeyStr = "2c6b29d5545e5804b7fb7e65404782e57dddea84cbcec9b2586add6403808965";
var BobFilEthPrivateKey = new (0, _crypto.SECP256K1PrivateKey)(_encoding.decodeHex.call(void 0, BobFilEthPrivateKeyStr));
var FILECOIN_PREFIX_F4_ETHEREUM = "f410";
var FILECOIN_PREFIX_T4_ETHEREUM = "t410";












exports.AliceFilStr = AliceFilStr; exports.AliceFilEthAddressStr = AliceFilEthAddressStr; exports.AliceFilEthPrivateKeyStr = AliceFilEthPrivateKeyStr; exports.AliceFilEthPrivateKey = AliceFilEthPrivateKey; exports.BobFilAddressStr = BobFilAddressStr; exports.BobFilEthAddressStr = BobFilEthAddressStr; exports.BobFilEthPrivateKeyStr = BobFilEthPrivateKeyStr; exports.BobFilEthPrivateKey = BobFilEthPrivateKey; exports.FILECOIN_PREFIX_F4_ETHEREUM = FILECOIN_PREFIX_F4_ETHEREUM; exports.FILECOIN_PREFIX_T4_ETHEREUM = FILECOIN_PREFIX_T4_ETHEREUM;
