// @mailchain/addressing/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkQV4RXTX4js = require('./chunk-QV4RXTX4.js');


var _chunkNKVNNBKEjs = require('./chunk-NKVNNBKE.js');


var _chunk6IENTSWHjs = require('./chunk-6IENTSWH.js');
require('./chunk-DFQJPZQW.js');
require('./chunk-GHT3H4ZF.js');
require('./chunk-UV736ZKK.js');
require('./chunk-ZAAEECZB.js');


var _chunkD5PLNGK2js = require('./chunk-D5PLNGK2.js');








var _chunkKULIOUHZjs = require('./chunk-KULIOUHZ.js');




var _chunkE6OAXMPTjs = require('./chunk-E6OAXMPT.js');


var _chunkOS5AHDT5js = require('./chunk-OS5AHDT5.js');


var _chunkZVESC2VGjs = require('./chunk-ZVESC2VG.js');



var _chunkOPDUNZOIjs = require('./chunk-OPDUNZOI.js');
require('./chunk-L7UXQ4P2.js');
require('./chunk-BAZMUWY4.js');
require('./chunk-PR4HBQYQ.js');
require('./chunk-PV3ICE6E.js');
require('./chunk-XESCJ3QG.js');
require('./chunk-P66WQYVA.js');
require('./chunk-7QWM3DS3.js');
require('./chunk-O73NCH4N.js');
require('./chunk-QBWJRO3P.js');




var _chunkAXBYMAV7js = require('./chunk-AXBYMAV7.js');
require('./chunk-4WTCQFUO.js');
require('./chunk-MFGUARH3.js');










var _chunk7K2RPUW3js = require('./chunk-7K2RPUW3.js');


var _chunk6IKXH2YOjs = require('./chunk-6IKXH2YO.js');
require('./chunk-XXPGZHWZ.js');

// src/test.constants.ts
var EthereumAlice = new Uint8Array([
  213,
  171,
  76,
  227,
  96,
  92,
  213,
  144,
  219,
  96,
  155,
  107,
  92,
  137,
  1,
  253,
  178,
  239,
  127,
  230
]);
var EthereumBob = new Uint8Array([
  146,
  216,
  241,
  2,
  72,
  198,
  163,
  149,
  60,
  195,
  105,
  42,
  137,
  70,
  85,
  173,
  5,
  214,
  30,
  251
]);


































exports.ALGORAND = _chunk7K2RPUW3js.ALGORAND; exports.ALL_PROTOCOLS = _chunk7K2RPUW3js.ALL_PROTOCOLS; exports.BadlyFormattedAddressError = _chunkOPDUNZOIjs.BadlyFormattedAddressError; exports.ETHEREUM = _chunk7K2RPUW3js.ETHEREUM; exports.EthereumAlice = EthereumAlice; exports.EthereumBob = EthereumBob; exports.IdentityProviderAddressInvalidError = _chunkOPDUNZOIjs.IdentityProviderAddressInvalidError; exports.MAILCHAIN = _chunk7K2RPUW3js.MAILCHAIN; exports.NEAR = _chunk7K2RPUW3js.NEAR; exports.ProtocolNotSupportedError = _chunk6IKXH2YOjs.ProtocolNotSupportedError; exports.SOLANA = _chunk7K2RPUW3js.SOLANA; exports.SUBSTRATE = _chunk7K2RPUW3js.SUBSTRATE; exports.TEZOS = _chunk7K2RPUW3js.TEZOS; exports.addressFromPublicKey = _chunkD5PLNGK2js.addressFromPublicKey; exports.checkAddressForErrors = _chunkZVESC2VGjs.checkAddressForErrors; exports.createNameServiceAddress = _chunkE6OAXMPTjs.createNameServiceAddress; exports.createWalletAddress = _chunkQV4RXTX4js.createWalletAddress; exports.decodeAddressByProtocol = _chunkAXBYMAV7js.decodeAddressByProtocol; exports.encodeAddressByProtocol = _chunkAXBYMAV7js.encodeAddressByProtocol; exports.encodingByProtocol = _chunkAXBYMAV7js.encodingByProtocol; exports.formatAddress = _chunk6IENTSWHjs.formatAddress; exports.isBlockchainProtocolEnabled = _chunk7K2RPUW3js.isBlockchainProtocolEnabled; exports.isEthereumAddress = _chunkKULIOUHZjs.isEthereumAddress; exports.isFilecoinAddress = _chunkKULIOUHZjs.isFilecoinAddress; exports.isMailchainAccountAddress = _chunkKULIOUHZjs.isMailchainAccountAddress; exports.isNameServiceAddress = _chunkE6OAXMPTjs.isNameServiceAddress; exports.isNearImplicitAccount = _chunkKULIOUHZjs.isNearImplicitAccount; exports.isSameAddress = _chunkNKVNNBKEjs.isSameAddress; exports.isSolanaAddress = _chunkKULIOUHZjs.isSolanaAddress; exports.isTezosAddress = _chunkKULIOUHZjs.isTezosAddress; exports.isTokenAddress = _chunkKULIOUHZjs.isTokenAddress; exports.parseNameServiceAddress = _chunkE6OAXMPTjs.parseNameServiceAddress; exports.parseWalletAddress = _chunkOS5AHDT5js.parseWalletAddress;
