// @mailchain/signatures/chunk-WVJPHFO2.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkXXPGZHWZjs = require('./chunk-XXPGZHWZ.js');

// src/errors.ts
var ProtocolIsEmptyError = class extends Error {
  constructor() {
    super("protocol is empty");
  }
};
var AddressIsEmptyError = class extends Error {
  constructor() {
    super("address is empty");
  }
};
var MessagingKeyVerificationError = class extends Error {
  constructor() {
    super("Messaging key validation failed and is not useable for this address.");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "type", "messaging_key_validation_failed");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "docs", "https://docs.mailchain.com/developer/errors/codes#messaging_key_validation_failed");
  }
};
var ProvidedMessagingKeyIncorrectError = class extends Error {
  constructor(type) {
    super(`Provided messaging key does not match. Use the latest messaging key owned by ${type}.`);
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "type", "provided_messaging_key_incorrect");
    _chunkXXPGZHWZjs.__publicField.call(void 0, this, "docs", "https://docs.mailchain.com/developer/errors/codes#provided_messaging_key_incorrect");
  }
};
var AddressMustBeProtocolAddressError = class extends Error {
  constructor() {
    super("address must be a protocol address");
  }
};
var PublicKeyNotFoundError = class extends Error {
  constructor() {
    super("Mailchain public key not found");
  }
};








exports.ProtocolIsEmptyError = ProtocolIsEmptyError; exports.AddressIsEmptyError = AddressIsEmptyError; exports.MessagingKeyVerificationError = MessagingKeyVerificationError; exports.ProvidedMessagingKeyIncorrectError = ProvidedMessagingKeyIncorrectError; exports.AddressMustBeProtocolAddressError = AddressMustBeProtocolAddressError; exports.PublicKeyNotFoundError = PublicKeyNotFoundError;
