// @mailchain/signatures/index.mjs
import {
  verify
} from "./chunk-ZRFPCLKK.mjs";
import "./chunk-RICUBO5V.mjs";
import {
  mailchainDeliveryConfirmationMessage,
  signMailchainDeliveryConfirmation
} from "./chunk-CL3VSPB7.mjs";
import {
  mailchainProvidedMessagingKeyMessage,
  signMailchainProvidedMessagingKey,
  verifyMailchainProvidedMessagingKey
} from "./chunk-HZRXR7TV.mjs";
import {
  mailchainPasswordResetMessage,
  signMailchainPasswordReset,
  verifyMailchainPasswordReset
} from "./chunk-SCULSYUL.mjs";
import {
  signMailchainUsername,
  verifyMailchainUsername
} from "./chunk-IY4ZSTBW.mjs";
import {
  createMailerProof,
  createMailerProofSigningData,
  signMailerProofParams,
  verifyMailerProof
} from "./chunk-GVD4MK53.mjs";
import {
  signRawEd25519,
  verifyRawEd25519
} from "./chunk-RNDAINMT.mjs";
import {
  KindEthereumPersonalMessage,
  KindMailchainUsernameIdentityKey,
  KindRawED25519,
  KindTezos
} from "./chunk-XRFLHMTB.mjs";
import {
  AddressIsEmptyError,
  AddressMustBeProtocolAddressError,
  MessagingKeyVerificationError,
  ProtocolIsEmptyError,
  ProvidedMessagingKeyIncorrectError,
  PublicKeyNotFoundError
} from "./chunk-QFIG37GA.mjs";
import {
  getMessageHash,
  signEthereumPersonalMessage,
  verifyEthereumPersonalMessage
} from "./chunk-S45KIVEU.mjs";
import "./chunk-NHABU752.mjs";
export {
  AddressIsEmptyError,
  AddressMustBeProtocolAddressError,
  KindEthereumPersonalMessage,
  KindMailchainUsernameIdentityKey,
  KindRawED25519,
  KindTezos,
  MessagingKeyVerificationError,
  ProtocolIsEmptyError,
  ProvidedMessagingKeyIncorrectError,
  PublicKeyNotFoundError,
  createMailerProof,
  createMailerProofSigningData,
  getMessageHash,
  mailchainDeliveryConfirmationMessage,
  mailchainPasswordResetMessage,
  mailchainProvidedMessagingKeyMessage,
  signEthereumPersonalMessage,
  signMailchainDeliveryConfirmation,
  signMailchainPasswordReset,
  signMailchainProvidedMessagingKey,
  signMailchainUsername,
  signMailerProofParams,
  signRawEd25519,
  verify,
  verifyEthereumPersonalMessage,
  verifyMailchainPasswordReset,
  verifyMailchainProvidedMessagingKey,
  verifyMailchainUsername,
  verifyMailerProof,
  verifyRawEd25519
};
